In [9]:
import pandas as pd 
import os
from tqdm import tqdm
root_path = "csv/traffic"
drop = ['samples_below_100pct_ff', 'samples_below_95pct_ff', 'samples_below_90pct_ff', 'samples_below_80pct_ff.1',
          'samples_below_80pct_ff', 'samples_below_75pct_ff', 'samples_below_70pct_ff', 'samples_below_65pct_ff',
          'samples_below_60pct_ff', 'samples_below_55pct_ff', 'samples_below_50pct_ff', 'samples_below_45pct_ff',
          'samples_below_40pct_ff', 'samples_below_35pct_ff', 'samples_below_30pct_ff', 'samples_below_25pct_ff',
          'samples_below_20pct_ff', 'samples_below_15pct_ff', 'samples_below_10pct_ff', 'samples_below_5pct_ff']


files = os.listdir(root_path)
for f in tqdm(files):
    df = pd.read_csv('{}/{}'.format(root_path, f))
    try:
        df.drop(columns=drop, inplace=True)
        df.to_csv('{}/{}'.format(root_path, f))
    except ValueError:
        print('Already droped')  ...................................-      
    




  0%|                                                  | 0/194 [00:00<?, ?it/s]
Exception in thread Thread-7:
Traceback (most recent call last):
  File "C:\Users\arthur\Anaconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Users\arthur\Anaconda3\lib\site-packages\tqdm\_tqdm.py", line 148, in run
    for instance in self.tqdm_cls._instances:
  File "C:\Users\arthur\Anaconda3\lib\_weakrefset.py", line 60, in __iter__
    for itemref in self.data:
RuntimeError: Set changed size during iteration



Already droped


  1%|▏                                      | 1/194 [01:43<5:31:31, 103.07s/it]

Already droped


100%|█████████████████████████████████████| 194/194 [6:08:57<00:00, 114.11s/it]


In [18]:
for col in df:
    print(df[col].describe())
    print('----------')


count    2.017846e+06
mean     1.008922e+06
std      5.825021e+05
min      0.000000e+00
25%      5.044612e+05
50%      1.008922e+06
75%      1.513384e+06
max      2.017845e+06
Name: Unnamed: 0, dtype: float64
----------
count                 2017846
unique                    288
top       03/03/2018 21:10:00
freq                     9576
Name: utc_time_id, dtype: object
----------
count     2017846
unique          2
top           tmc
freq      1609313
Name: source_ref, dtype: object
----------
count       2017846
unique        30141
top       106+23194
freq            288
Name: source_id, dtype: object
----------
count       2017846
unique            1
top       hereA0106
freq        2017846
Name: feed_id, dtype: object
----------
count    2.017846e+06
mean     9.866922e-01
std      1.145891e-01
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      1.000000e+00
Name: primary_link_source_flag, dtype: float64
----------
count    2.017846e+06
mea

## Column analyzing
Analyzing each columns in the data frame, it turned clear that only 3 columns is from our interrest: **utc_time_id, source_id and avg_speed**.



## source_id

The column source_id give us 30141 unique values, refering to 30141 diferrent measurement points through the city in question(San Francisco), futher our dataframe will have the dimensionalite of ** [N, 30141]**, where N is the number of samples, and each column contains the avg_speed of a specific measurement point 

In [19]:
df.head()

Unnamed: 0          utc_time_id source_ref  source_id    feed_id  \
0           0  03/03/2018 07:30:00        tmc  106-13503  hereA0106   
1           1  03/03/2018 07:35:00        tmc  106-13503  hereA0106   
2           2  03/03/2018 07:40:00        tmc  106-13503  hereA0106   
3           3  03/03/2018 07:45:00        tmc  106-13503  hereA0106   
4           4  03/03/2018 07:50:00        tmc  106-13503  hereA0106   

   primary_link_source_flag  samples  avg_speed  avg_flow  avg_occ  \
0                         1        5       49.8       NaN      NaN   
1                         1        5       50.0       NaN      NaN   
2                         1        5       50.0       NaN      NaN   
3                         1        5       50.0       NaN      NaN   
4                         1        5       50.0       NaN      NaN   

   avg_freeflow_speed  avg_travel_time  high_quality_samples  
0                50.0             0.96                     0  
1                50.0             0.96                     0  
2                50.0             0.96                     0  
3                50.0             0.96                     0  
4                50.0             0.96                     0

## Time column
The time utc_time_column is in string format, not in datetime format, it is needed to convert for easier use in the future

In [21]:

for f in tqdm(files):
    df = pd.read_csv('{}/{}'.format(root_path, f))
    df = df[['utc_time_id', 'source_id', 'avg_speed']]
    df['utc_time_id'] = pd.to_datetime(df['utc_time_id'], infer_datetime_format=True)
    df.to_csv('{}/{}'.format(root_path, f), index=False)
    


  0%|                                                  | 0/194 [00:00<?, ?it/s]
Exception in thread Thread-8:
Traceback (most recent call last):
  File "C:\Users\arthur\Anaconda3\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Users\arthur\Anaconda3\lib\site-packages\tqdm\_tqdm.py", line 148, in run
    for instance in self.tqdm_cls._instances:
  File "C:\Users\arthur\Anaconda3\lib\_weakrefset.py", line 60, in __iter__
    for itemref in self.data:
RuntimeError: Set changed size during iteration

100%|█████████████████████████████████████| 194/194 [5:49:40<00:00, 108.15s/it]


## File manager

Due the high volume of information that needs to be process, we will create a way to process the data in multiprocess and in diferents computers 

In [1]:
import os
import pickle as pkl
from queue import Queue
import random

class FileManager(object):
    def __init__(self, directory = 'csv/traffic', path = 'C:/Users/arthur/Dropbox/TrafficFlow/lista.pkl'):

        self.files = os.listdir(directory)
        self.done = []
        self.processing = []
        self.path = path
        
    def get_item(self):
        x = self.files.pop(random.randint(0,len(self.files)-1))
        self.processing.append(x)
        save_pkl(self, self.path)
        return x
    def finnish(self, x):
        self.processing.remove(x)
        self.done.append(x)
        save_pkl(self, self.path)
        
    def set_path(self, path):
        self.path = path
        
    def __str__(self):
        msg = '{} itens to be process \n {} itens processed'.format(len(self.files), len(self.processing) + len(self.done))
        return msg
    
def save_pkl(obj, path = 'C:/Users/arthur/Dropbox/TrafficFlow/lista.pkl'):
    pkl.dump(obj, open(path, 'wb'))
    
def load_pkl(path = 'C:/Users/arthur/Dropbox/TrafficFlow/lista.pkl'):
    obj = pkl.load(open(path, 'rb'))
    if len(obj.processing) > 0:
        print('The process was stoped with running process')
        print('Restoring process to process queue')
        obj.files += obj.processing
        obj.processing = []
    return obj
 


In [1]:
dropdox_path = 'C:/Users/arthur/Dropbox/TrafficFlow/lista.pkl'
dir_path = 'csv/traffic'


In [2]:
def get_date(file):
    name, ext = file.rsplit('.')
    return name[-10:]



In [3]:
import pandas as pd
from tqdm import tqdm

def preprocess(file):
    df = pd.read_csv(file, dtype={'source_id':'category'})
    path = 'csv/traffic/{}.csv'.format(get_date(file))
    print(path)
    df_res = df.pivot(index='utc_time_id', columns='source_id', values='avg_speed')
    df_res.to_csv(path, index=True)
    return file
    



In [4]:

class Task(object):
    def __init__(self, file, user, status):
        self.file = file
        self.user = user
        self.status = status 
        

    def get_file(self):
        self.status = 'running'
        return self.file
        
    def __str__(self):
        return '{} - {}'.format(self.user, self.file)
        
def get_status(obj):
    arthur_count = 0
    zaroni_count = 0
    tony_count = 0
    total = len(obj)
    for task in obj:
        if task.status == 'done':
            if task.user == 'arthur':
                arthur_count += 1
            elif task.user == 'zaroni':
                zaroni_count += 1
            else:
                tony_count += 1
    print('Arthur converted {}'.format(arthur_count))
    print('Zaroni converted {}'.format(zaroni_count))
    print('Tony converted {}'.format(tony_count))
    print('{} still missing'.format(total - arthur_count - zaroni_count - tony_count))
    print('Progress: {}/{}'.format((arthur_count+zaroni_count+tony_count),total))

In [5]:
import os 
import pickle as pkl 
direc = os.listdir(dir_path)

create = False

if create:
    len_ = len(direc)
    lista = []
    for i, element in enumerate(direc):
        if i < len_/3:
            lista.append(Task('{}/{}'.format(dir_path, element), 'arthur', 'raw'))
        elif i < len_*2/3:
            lista.append(Task(element, 'Zaroni', 'raw'))
        else:
            lista.append(Task(element, 'Tony', 'raw'))

    pkl.dump(lista, open(dropdox_path, 'wb'))
else:
    lista = pkl.load(open(dropdox_path, 'rb'))


In [8]:
user = 'arthur'


for l in tqdm(lista):
    if l.status == 'raw':
        files_converted = preprocess(l.get_file())
        l.status = 'done'
        pkl.dump(lista, open(dropdox_path, 'wb'))

  0%|                                                  | 0/154 [00:00<?, ?it/s]

csv/traffic/2017_12_11.csv


 47%|███████████████████▏                     | 72/154 [00:49<00:56,  1.44it/s]

csv/traffic/2017_12_12.csv


 47%|███████████████████▍                     | 73/154 [01:44<01:55,  1.42s/it]

csv/traffic/2017_12_13.csv


 48%|███████████████████▋                     | 74/154 [02:37<02:50,  2.13s/it]

csv/traffic/2017_12_14.csv


 49%|███████████████████▉                     | 75/154 [03:29<03:40,  2.79s/it]

csv/traffic/2017_12_15.csv


 49%|████████████████████▏                    | 76/154 [04:24<04:31,  3.48s/it]

csv/traffic/2017_12_16.csv


 50%|████████████████████▌                    | 77/154 [04:59<04:59,  3.88s/it]

csv/traffic/2017_12_17.csv


 51%|████████████████████▊                    | 78/154 [05:27<05:18,  4.20s/it]

csv/traffic/2017_12_18.csv


 51%|█████████████████████                    | 79/154 [06:16<05:57,  4.77s/it]

csv/traffic/2017_12_19.csv


 52%|█████████████████████▎                   | 80/154 [07:10<06:38,  5.38s/it]

csv/traffic/2017_12_20.csv


 53%|█████████████████████▌                   | 81/154 [08:02<07:14,  5.96s/it]

csv/traffic/2017_12_21.csv


 53%|█████████████████████▊                   | 82/154 [08:50<07:45,  6.46s/it]

csv/traffic/2017_12_22.csv


 54%|██████████████████████                   | 83/154 [09:37<08:13,  6.95s/it]

csv/traffic/2017_12_23.csv


 55%|██████████████████████▎                  | 84/154 [10:08<08:27,  7.25s/it]

csv/traffic/2017_12_24.csv


 55%|██████████████████████▋                  | 85/154 [10:31<08:32,  7.43s/it]

csv/traffic/2017_12_25.csv


 56%|██████████████████████▉                  | 86/154 [10:48<08:32,  7.54s/it]

csv/traffic/2017_12_26.csv


 56%|███████████████████████▏                 | 87/154 [11:29<08:50,  7.92s/it]

csv/traffic/2017_12_27.csv


 57%|███████████████████████▍                 | 88/154 [12:12<09:09,  8.33s/it]

csv/traffic/2017_12_28.csv


 58%|███████████████████████▋                 | 89/154 [12:56<09:27,  8.72s/it]

csv/traffic/2017_12_29.csv


 58%|███████████████████████▉                 | 90/154 [13:43<09:45,  9.15s/it]

csv/traffic/2017_12_30.csv


 59%|████████████████████████▏                | 91/154 [14:18<09:54,  9.44s/it]

csv/traffic/2017_12_31.csv


 60%|████████████████████████▍                | 92/154 [14:41<09:53,  9.58s/it]

csv/traffic/2018_01_01.csv


 60%|████████████████████████▊                | 93/154 [15:04<09:53,  9.72s/it]

csv/traffic/2018_01_02.csv


 61%|█████████████████████████                | 94/154 [16:03<10:14, 10.25s/it]

csv/traffic/2018_01_03.csv


 62%|█████████████████████████▎               | 95/154 [16:53<10:29, 10.67s/it]

csv/traffic/2018_01_04.csv


 62%|█████████████████████████▌               | 96/154 [17:39<10:40, 11.04s/it]

csv/traffic/2018_01_05.csv


 63%|█████████████████████████▊               | 97/154 [18:29<10:52, 11.44s/it]

csv/traffic/2018_01_06.csv


 64%|██████████████████████████               | 98/154 [19:05<10:54, 11.69s/it]

csv/traffic/2018_01_07.csv


 64%|██████████████████████████▎              | 99/154 [19:29<10:49, 11.82s/it]

csv/traffic/2018_01_08.csv


 65%|█████████████████████████▉              | 100/154 [20:19<10:58, 12.20s/it]

csv/traffic/2018_01_09.csv


 66%|██████████████████████████▏             | 101/154 [21:17<11:10, 12.65s/it]

csv/traffic/2018_01_10.csv


 66%|██████████████████████████▍             | 102/154 [22:15<11:21, 13.10s/it]

csv/traffic/2018_01_11.csv


 67%|██████████████████████████▊             | 103/154 [23:09<11:27, 13.49s/it]

csv/traffic/2018_01_12.csv


 68%|███████████████████████████             | 104/154 [24:02<11:33, 13.87s/it]

csv/traffic/2018_01_13.csv


 68%|███████████████████████████▎            | 105/154 [24:42<11:31, 14.12s/it]

csv/traffic/2018_01_14.csv


 69%|███████████████████████████▌            | 106/154 [25:16<11:26, 14.31s/it]

csv/traffic/2018_01_15.csv


 69%|███████████████████████████▊            | 107/154 [26:06<11:27, 14.64s/it]

csv/traffic/2018_01_16.csv


 70%|████████████████████████████            | 108/154 [26:54<11:27, 14.95s/it]

csv/traffic/2018_01_17.csv


 71%|████████████████████████████▎           | 109/154 [27:45<11:27, 15.28s/it]

csv/traffic/2018_01_18.csv


 71%|████████████████████████████▌           | 110/154 [28:35<11:26, 15.59s/it]

csv/traffic/2018_01_19.csv


 72%|████████████████████████████▊           | 111/154 [29:26<11:24, 15.91s/it]

csv/traffic/2018_01_20.csv


 73%|█████████████████████████████           | 112/154 [29:41<11:08, 15.91s/it]

csv/traffic/2018_01_21.csv


 73%|█████████████████████████████▎          | 113/154 [30:08<10:56, 16.01s/it]

csv/traffic/2018_01_22.csv


 74%|█████████████████████████████▌          | 114/154 [31:06<10:55, 16.38s/it]

csv/traffic/2018_01_23.csv


 75%|█████████████████████████████▊          | 115/154 [31:58<10:50, 16.69s/it]

csv/traffic/2018_01_24.csv


 75%|██████████████████████████████▏         | 116/154 [32:47<10:44, 16.96s/it]

csv/traffic/2018_01_25.csv


 76%|██████████████████████████████▍         | 117/154 [33:36<10:37, 17.23s/it]

csv/traffic/2018_01_26.csv


 77%|██████████████████████████████▋         | 118/154 [34:39<10:34, 17.62s/it]

csv/traffic/2018_01_27.csv


 77%|██████████████████████████████▉         | 119/154 [35:22<10:24, 17.83s/it]

csv/traffic/2018_01_28.csv


 78%|███████████████████████████████▏        | 120/154 [35:55<10:10, 17.96s/it]

csv/traffic/2018_01_29.csv


 79%|███████████████████████████████▍        | 121/154 [36:52<10:03, 18.28s/it]

csv/traffic/2018_01_30.csv


 79%|███████████████████████████████▋        | 122/154 [37:51<09:55, 18.62s/it]

csv/traffic/2018_01_31.csv


 80%|███████████████████████████████▉        | 123/154 [38:48<09:46, 18.93s/it]

csv/traffic/2018_02_01.csv


 81%|████████████████████████████████▏       | 124/154 [39:49<09:38, 19.27s/it]

csv/traffic/2018_02_02.csv


 81%|████████████████████████████████▍       | 125/154 [40:48<09:28, 19.59s/it]

csv/traffic/2018_02_03.csv


 82%|████████████████████████████████▋       | 126/154 [41:29<09:13, 19.76s/it]

csv/traffic/2018_02_04.csv


 82%|████████████████████████████████▉       | 127/154 [41:59<08:55, 19.84s/it]

csv/traffic/2018_02_05.csv


 83%|█████████████████████████████████▏      | 128/154 [42:58<08:43, 20.14s/it]

csv/traffic/2018_02_06.csv


 84%|█████████████████████████████████▌      | 129/154 [43:54<08:30, 20.42s/it]

csv/traffic/2018_02_07.csv


 84%|█████████████████████████████████▊      | 130/154 [44:42<08:15, 20.63s/it]

csv/traffic/2018_02_08.csv


 85%|██████████████████████████████████      | 131/154 [45:32<07:59, 20.86s/it]

csv/traffic/2018_02_09.csv


 86%|██████████████████████████████████▎     | 132/154 [46:27<07:44, 21.12s/it]

csv/traffic/2018_02_10.csv


 86%|██████████████████████████████████▌     | 133/154 [47:00<07:25, 21.20s/it]

csv/traffic/2018_02_11.csv


 87%|██████████████████████████████████▊     | 134/154 [47:24<07:04, 21.23s/it]

csv/traffic/2018_02_12.csv


 88%|███████████████████████████████████     | 135/154 [48:10<06:46, 21.41s/it]

csv/traffic/2018_02_13.csv


 88%|███████████████████████████████████▎    | 136/154 [49:01<06:29, 21.63s/it]

csv/traffic/2018_02_14.csv


 89%|███████████████████████████████████▌    | 137/154 [49:50<06:11, 21.83s/it]

csv/traffic/2018_02_15.csv


 90%|███████████████████████████████████▊    | 138/154 [50:40<05:52, 22.04s/it]

csv/traffic/2018_02_16.csv


 90%|████████████████████████████████████    | 139/154 [51:29<05:33, 22.23s/it]

csv/traffic/2018_02_17.csv


 91%|████████████████████████████████████▎   | 140/154 [52:02<05:12, 22.31s/it]

csv/traffic/2018_02_18.csv


 92%|████████████████████████████████████▌   | 141/154 [52:28<04:50, 22.33s/it]

csv/traffic/2018_02_19.csv


 92%|████████████████████████████████████▉   | 142/154 [53:11<04:29, 22.47s/it]

csv/traffic/2018_02_20.csv


 93%|█████████████████████████████████████▏  | 143/154 [53:59<04:09, 22.65s/it]

csv/traffic/2018_02_21.csv


 94%|█████████████████████████████████████▍  | 144/154 [54:46<03:48, 22.82s/it]

csv/traffic/2018_02_22.csv


 94%|█████████████████████████████████████▋  | 145/154 [55:37<03:27, 23.02s/it]

csv/traffic/2018_02_23.csv


 95%|█████████████████████████████████████▉  | 146/154 [56:28<03:05, 23.21s/it]

csv/traffic/2018_02_24.csv


 95%|██████████████████████████████████████▏ | 147/154 [56:51<02:42, 23.21s/it]

csv/traffic/2018_02_25.csv


 96%|██████████████████████████████████████▍ | 148/154 [57:17<02:19, 23.22s/it]

csv/traffic/2018_02_26.csv


 97%|██████████████████████████████████████▋ | 149/154 [58:04<01:56, 23.38s/it]

csv/traffic/2018_02_27.csv


 97%|██████████████████████████████████████▉ | 150/154 [58:55<01:34, 23.57s/it]

csv/traffic/2018_02_28.csv


 98%|███████████████████████████████████████▏| 151/154 [59:44<01:11, 23.74s/it]

csv/traffic/2018_03_01.csv


 99%|█████████████████████████████████████▌| 152/154 [1:00:30<00:47, 23.89s/it]

csv/traffic/2018_03_02.csv


 99%|█████████████████████████████████████▊| 153/154 [1:01:18<00:24, 24.04s/it]

csv/traffic/2018_03_03.csv


100%|██████████████████████████████████████| 154/154 [1:01:53<00:00, 24.12s/it]


In [11]:
for i, l in enumerate(lista):
    print(i, l.file, l.status, l.user)

0 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_01.csv done arthur
1 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_02.csv done arthur
2 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_03.csv done arthur
3 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_04.csv done arthur
4 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_05.csv done arthur
5 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_06.csv done arthur
6 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_07.csv done arthur
7 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_08.csv done arthur
8 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_09.csv done arthur
9 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_10.csv done arthur
10 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_11.csv done arthur
11 csv/traffic/contracted-Caltrans_text_gn_link_5min_2017_10_12.csv done arthur
12 csv/traffic/contracted-Caltrans_text_gn_link_5m

In [7]:
df = pd.read_csv(file, dtype={'source_id':'category'})
path = 'csv/traffic/{}.csv'.format(get_date(file))
print(path)
df_res = df.pivot(index='utc_time_id', columns='source_id', values='avg_speed')





'csv/traffic'

'csv/traffic'